src/lib.rs

```diff
@@ -1,8 +1,530 @@
+    ///
+    /// 1. A plain trait name will implement the trait for *all* objects generated from
+    ///    the custom resource definition: `--derive PartialEq`
+    ///
+    /// 2. Constraining the derivation to a singular struct or enum:
+    ///    `--derive IssuerAcmeSolversDns01CnameStrategy=PartialEq`
+    ///
+    /// 3. Constraining the derivation to only structs (@struct), enums (@enum) or *unit-only* enums (@enum:simple),
+    ///    meaning enums where no variants are tuple or structs:
+    ///    `--derive @struct=PartialEq`, `--derive @enum=PartialEq`, `--derive @enum:simple=PartialEq`
+    ///
+    /// See also: https://doc.rust-lang.org/reference/items/enumerations.html
+    #[cfg_attr(feature = "cli", arg(
+        short = 'D',
+        long = "derive",
+        value_parser = Derive::from_str,
+    ))]
+    #[builder(via_mutators(init = Default::default()))]
+    pub derive_traits: Vec<Derive>,
+
+    /// Elide the following containers from the output
+    ///
+    /// This allows manual customization of structs from the output without having to remove it from
+    /// the output first. Takes precise generated struct names.
+    #[cfg_attr(feature = "cli", arg(long, short = 'e'))]
+    pub elide: Vec<String>,
+
+    /// Relaxed interpretation
+    ///
+    /// This allows certain invalid openapi specs to be interpreted as arbitrary objects as used by
+    /// argo workflows, for example.
+    #[cfg_attr(feature = "cli", arg(long))]
+    pub relaxed: bool,
+
+    /// Disable standardized Condition API
+    ///
+    /// By default, kopium detects Condition objects and uses a standard
+    /// Condition API from k8s_openapi instead of generating a custom definition.
+    #[cfg_attr(feature = "cli", arg(long))]
+    pub no_condition: bool,
+
+    /// Disable standardised ObjectReference API
+    ///
+    /// By default, kopium detects ObjectReference objects and uses a standard
+    /// ObjectReference from k8s_openapi instead of generating a custom definition.
+    #[cfg_attr(feature = "cli", arg(long))]
+    pub no_object_reference: bool,
+
+    /// Type used to represent maps via `additionalProperties`
+    #[builder(setter(into))]
+    #[cfg_attr(feature = "cli", arg(long, value_enum, default_value_t))]
+    pub map_type: MapType,
+
+    /// Automatically removes `#[derive(Default)]` from structs that contain fields for
+    /// which a default cannot be automatically derived.
+    ///
+    /// This option only has an effect if `--derive Default` is set.
+    #[cfg_attr(feature = "cli", arg(long))]
+    pub smart_derive_elision: bool,
+}
+
+impl Default for TypeGenerator {
+    fn default() -> Self {
+        Self::builder().build()
+    }
+}
+
+impl TypeGenerator {
+    pub async fn generate_rust_types_for(
+        &self,
+        crd: &CustomResourceDefinition,
+        args: Option<impl std::fmt::Display>,
+    ) -> anyhow::Result<String> {
+        use std::fmt::Write;
```

Comment:
this almost feels like an impl Display fmt fn now. maybe the crd and args could be lifted into the builder so that the final format function is equivalent to a display? happy to skip this for a this first PR